cmd/agentbasedinstaller/client/main.go

```diff
@@ -171,11 +185,13 @@ func registerCluster(ctx context.Context, log *log.Logger, bmInventory *client.A
 	modelsCluster, err := agentbasedinstaller.RegisterCluster(ctx, log, bmInventory, pullSecret,
 		RegisterOptions.ClusterDeploymentFile, RegisterOptions.AgentClusterInstallFile, RegisterOptions.ClusterImageSetFile, RegisterOptions.ReleaseImageMirror, RegisterOptions.OperatorInstallFile)
 	if err != nil {
+		deregisterCluster(ctx, log, bmInventory)
 		log.Fatal("Failed to register cluster with assisted-service: ", err)
 	}
 
 	err = agentbasedinstaller.RegisterExtraManifests(os.DirFS(RegisterOptions.ExtraManifests), ctx, log, bmInventory.Manifests, modelsCluster)
 	if err != nil {
+		deregisterCluster(ctx, log, bmInventory)
```

Comment:
In the case of a deterministic error like the one that led us to discovering the problem, this would be sufficient. But in general, recovering from errors in error-prone actions by performing more error-prone actions is... error-prone. I think we need to treat this like a distributed system, with the goal of it eventually converging to the desired state.